revolv/payments/signals.py

```diff
@@ -0,0 +1,146 @@
+from django.db.models import Sum, signals
+from django.dispatch import receiver
+from revolv.base.models import RevolvUserProfile
+from revolv.payments.models import (AdminReinvestment, AdminRepayment, Payment,
+                                    PaymentType, Repayment)
+from revolv.payments.utils import (NotEnoughFundingException,
+                                   ProjectNotCompleteException)
+
+
+@receiver(signals.pre_init, sender=AdminRepayment)
+def pre_init_admin_repayment(**kwargs):
+    """
+    Make sure that related project is indeed complete, else throw a
+    ProjectNotCompleteException and to disallow instantiation of an invalid
+    AdminRepayment.
+    """
+    init_kwargs = kwargs.get('kwargs')
+    if not init_kwargs:
+        raise NotEnoughFundingException()
+    project = init_kwargs.get('project')
+    if not project.is_completed:
+        raise ProjectNotCompleteException()
+
+
+@receiver(signals.post_save, sender=AdminRepayment)
+def post_save_admin_repayment(**kwargs):
+    """
+    When an AdminRepayment is saved, a Repayment is generated for all donors to
+    a project, each weighed by that donor's proportion of the contribution to
+    the project.
+    """
+    instance = kwargs.get('instance')
+    for donor in instance.project.donors.all():
+        repayment = Repayment(user=donor,
+                              project=instance.project,
+                              admin_repayment=instance,
+                              amount=(instance.project.proportion_donated(donor) * instance.amount),
+                              )
+        repayment.save()
+
+
+@receiver(signals.pre_init, sender=AdminReinvestment)
+def pre_init_admin_reinvestment(**kwargs):
+    """
+    Raises a NotEnoughFundingException before __init__ if there are not enough
+    funds for this AdminReinvestment.
+    """
+    init_kwargs = kwargs.get('kwargs')
+    if not init_kwargs:
+        raise NotEnoughFundingException()
+    invest_amount = init_kwargs.get('amount') or 0.0
+
+    global_repay_amount = AdminRepayment.objects.aggregate(
+        Sum('amount')
+    )['amount__sum'] or 0.0
+    global_reinvest_amount = AdminReinvestment.objects.aggregate(
+        Sum('amount')
+    )['amount__sum'] or 0.0
+    global_reinvest_pool = global_repay_amount - global_reinvest_amount
+    if global_reinvest_pool - invest_amount < 0.0:
```

Comment:
Uh, pretty sure this is right. It's checking to see that we won't dip below `$0` if we make this reinvestment.